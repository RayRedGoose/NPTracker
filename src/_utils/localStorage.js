export const getItem = item => {
  return JSON.parse(localStorage.getItem(item))
}

export const addItem = (type, item) => {
  return localStorage.setItem(type, JSON.stringify(item))
}

export const addItemToAll = (path, item) => {
  const targetPath = (!getItem(path)) ? [ ] : [...getItem(path)]
  return localStorage.setItem(path, JSON.stringify([...targetPath, item]))
}
