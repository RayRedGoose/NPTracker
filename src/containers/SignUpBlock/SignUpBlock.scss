@import "../../variables.scss";
@import "../../index.scss";

.signup-container {
  @extend %container;
  .arrow-btn {
    cursor: pointer;
    position: absolute;
    top: 47%; left: 28vw;
    width: 50px;
    opacity: 0.1;
    &:hover {
      opacity: 1;
    }
    &:active {
      opacity: 0.3;
    }
  }
}

.inactive-block {
  &:before {
    content: '';
    background-color: $light-color;
    margin-left: 12vw;
    position: absolute;
    top: 50%;
    width: 25vw; height: 1px;
  }
  &:last-child {
    &:before {
      content: '';
      background-color: $light-color;
      margin-left: 2vw;
      position: absolute;
      top: 50%;
      width: 10vw; height: 1px;
    }
  }
}

.active-block {
  @extend %active-container;
}

.submit-block {
  display: flex;
  flex-direction: column;
  align-items: center;
  .image {
    width: 200px; height: 200px;
    border: 1px solid $accent-color-primary-dark;
    background-position:  center;
    background-size: cover;
  }
  p {
    text-transform: uppercase;
    font-size: 24px;
    text-align: center;
    margin-top: 40px;
  }
  .location {
    text-transform: capitalize;
    margin-top: 10px;
    &:before {
      content: "";
      background-image: url('../../assets/pin.svg');
      background-position: center;
      background-repeat: no-repeat;
      width: 20px; height: 24px;
      display: inline-block;
    }
  }
  div {
    display: flex;
    margin: 40px;
    button {
      margin: 0 10px;
      &:first-child {
        color: $light-color;
        @extend %bordered-btn;
      }
      &:last-child {
        @extend %accent-btn;
      }
    }
  }
}

.error-block {
  .error {
    @extend .error;
    margin: 0 0 40px;
    text-transform: none;
    font-size: 18px;
  }
  button {
    @extend %accent-btn;
  }
}
