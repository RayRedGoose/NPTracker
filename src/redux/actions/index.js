export const addProcessName = name => ({
  type: 'ADD_PROCESS_NAME',
  name
})

export const addProcessStep = step => ({
  type: 'ADD_PROCESS_STEP',
  step
})

export const increaseStep = () => ({
  type: 'INCREASE_STEP'
})

export const addUserInfo = info => ({
  type: 'ADD_USER_INFO',
  info
})

export const addParks = parks => ({
  type: 'ADD_PARKS',
  parks
})

export const changeActiveTab = id => ({
  type: 'CHANGE_ACTIVE_TAB',
  id
})
