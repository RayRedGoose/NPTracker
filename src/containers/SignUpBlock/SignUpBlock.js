import './SignUpBlock.scss'
import React, { Component, Fragment } from 'react'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import matryoshka from 'assets/matryoshka.svg'
import arrow from 'assets/send.svg'
import { addItem, addItemToAll } from '_utils/localStorage'
import { addUserInfo, increaseStep } from 'redux/actions'

export class SignUpBlock extends Component {
  constructor(props) {
    super(props)
    this.state = {
      type: '',
      data: {},
      error: ''
    }
  }

  componentDidMount() {
    const { process } = this.props
    if (!process) window.location = '/welcome'
    this.checkType()
  }

  componentDidUpdate(prevProps) {
    if (this.props.process.step !== prevProps.process.step) {
      this.checkType()
    }
  }

  checkType = () => {
    const { id, process } = this.props
    if (process.step < id) this.setState({type: 'inactive'})
    if (process.step === id) this.setState({type: 'active'})
    if (process.step > id) this.setState({type: 'done'})
  }

  createInputs = () => {
    const { inputs } = this.props

    return inputs.map(input => {
      const text = (input === 'password') ? 'password' : 'text'
      return (
        <Fragment key="inputs">
          <label
            key={'label-' + input}
            htmlFor={input}>
            {input.split('_').join(' ')}</label>
          <input
            key={'input-' + input}
            id={input}
            type={text}
            value={this.state.data[input]}
            onChange={this.handleChanges} />
        </Fragment>
      )
    })
  }

  handleChanges = (event) => {
    this.setState({error: ''})
    const obj = {}
    obj[event.target.id] = event.target.value
    this.setState({data: {...this.state.data, ...obj}})
  }

  updateUser = () => {
    this.props.increaseStep()
    this.props.addUserInfo(this.state.data)
  }

  submitUser = () => {
    const {user} = this.props
    addItem('user', user)
    addItemToAll('users', user)
    window.location = '/'
  }

  cancel = () => {
    window.location = '/window'
  }

  checkUserFullness = () => {
    const user = this.props.user ? this.props.user : {}
    return Object.keys(user).sort() !== ['city', 'image', 'last_name','name','password', 'username']
  }

  render() {
    const { id, user, process } = this.props

    const imageStyle = {
      height: `${200 - (id * 40)}px`,
      position: 'relative',
      zIndex: '3'
    }

    const bodyWithInputs = (
      <Fragment key='form'>
        <form>
          <h2>CREATE ACCOUNT</h2>
          {!this.state.error === '' && <p className="error-notification">this.state.error</p>}
          { this.createInputs() }
          <p>click arrow button to go to next step</p>
        </form>
        <img className="arrow-btn" src={arrow} alt="arrow" onClick={this.updateUser}/>
      </Fragment>
    )

    const bodyWithError = (
      <div key='error-block' className="submit-block error-block">
        <h2>Profile overview</h2>
        <p className="error">You don't pass correct data</p>
        <button onClick={this.cancel}>Cancel</button>
      </div>
    )

    const bodyWithSubmit = (this.checkUserFullness()) ? bodyWithError : (
      <div key='submit-block' className="submit-block">
        <h2>Profile overview</h2>
        <img src={user.image} alt="avatar" />
        <p>{user.name + ' ' + user.last_name}</p>
        <p className="location">{user.city}</p>
        <div>
          <button onClick={this.submitUser}>Create</button>
          <button onClick={this.cancel}>Cancel</button>
        </div>
      </div>
    )

    const checkBody = (process && process.step === 4) ? bodyWithSubmit : bodyWithInputs

    return (
      <section id={ id } className={`signup-container ${this.state.type}-block`}>
        { this.state.type === 'inactive' && <img style={imageStyle} src={matryoshka} alt="matryoshka"/>}
        { this.state.type === 'active' && checkBody }
      </section>
    )
  }
}

export const mapStateToProps = ({ process, user }) => ({
  process, user
})

export const mapDispatchToProps = dispatch => (
  bindActionCreators({
    addUserInfo,
    increaseStep
  }, dispatch)
)

export default connect(mapStateToProps, mapDispatchToProps)(SignUpBlock)
