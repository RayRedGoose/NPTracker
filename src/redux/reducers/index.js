import { combineReducers } from 'redux'
import { process } from './singleReducers/process'
import { user } from './singleReducers/user'

const rootReducer = combineReducers({
  process,
  user
})

export default rootReducer
