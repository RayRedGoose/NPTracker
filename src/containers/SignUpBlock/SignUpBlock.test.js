import React from 'react'
import { shallow } from 'enzyme'
import { SignUpBlock, mapStateToProps, mapDispatchToProps } from './SignUpBlock'
import { addItem, addItemToAll } from '_utils/localStorage'
import { addUserInfo, increaseStep } from 'redux/actions'

jest.mock('_utils/localStorage')

describe("SignUpBlock", () => {
  let block, instance

  const mockUser = {
    username: 'some-user',
    password: '1234',
    name: 'Ray',
    last_name: 'Zlou',
    city: 'Denver',
    image: 'some-image'
  }
  const addUserInfo = jest.fn()
  const increaseStep = jest.fn()

  const mockProps = {
    key: 1,
    id: 1,
    inputs: ['username'],
    process: {
      name: 'signup',
      step: 1
    },
    user: {},
    addUserInfo,
    increaseStep
  }

  beforeEach(() => {
    block = shallow(
      <SignUpBlock {...mockProps} />
    )

    instance = block.instance()
  })

  describe("Snapshots", () => {

    it("should match snapshot with all data passed correctly", () => {
      expect(block).toMatchSnapshot()
    })

    it("should match snapshot if type is inactive", () => {
      block.setState({type: 'inactive'})
      expect(block).toMatchSnapshot()
    })

    it("should match snapshot if type is done", () => {
      block.setState({type: 'done'})
      expect(block).toMatchSnapshot()
    })

    it("should match snapshot if type is active on last step of process and all user data was entered", () => {
      const lastBlock = shallow(
        <SignUpBlock {...mockProps}
          id={4}
          process={{name: 'signup', step: 4}}
          user={mockUser} />
      )

      expect(lastBlock).toMatchSnapshot()
    })

    it("should match snapshot with error if type is active on last step of process but no user data", () => {
      const lastBlock = shallow(
        <SignUpBlock {...mockProps}
          id={4}
          process={{name: 'signup', step: 4}} />
      )

      expect(lastBlock).toMatchSnapshot()
    })
  })

  describe("Active state", () => {

    it("should call createInputs method after rendering", () => {
      const spy = jest.spyOn(instance, 'createInputs').mockImplementation(jest.fn())
      instance.forceUpdate()
      expect(spy).toHaveBeenCalled()
    })

    it("should call updateUser method when arrow image is clicked", () => {
      const spy = jest.spyOn(instance, 'updateUser').mockImplementation(jest.fn())
      instance.forceUpdate()

      block.find('img').simulate('click')

      expect(spy).toHaveBeenCalled()
    })

    it("should call handleChanges method when input is changed", () => {
      const spy = jest.spyOn(instance, 'handleChanges').mockImplementation(jest.fn())
      instance.forceUpdate()

      const mockEvent = {
        target: {
          id: 'username',
          value: 'user'
        }
      }

      block.find('input').simulate('change', mockEvent)

      expect(spy).toHaveBeenCalledWith(mockEvent)
    })

    describe("Last block", () => {
      let lastBlock, newInstance

      beforeEach(() => {
        lastBlock = shallow(
          <SignUpBlock {...mockProps}
            id={4}
            process={{name: 'signup', step: 4}}
            user={mockUser} />
        )

        newInstance = lastBlock.instance()
      })

      it("should call submitUser method when first button is clicked if it's lastBlock with correct user data", () => {
        const spy = jest.spyOn(newInstance, 'submitUser').mockImplementation(jest.fn())
        newInstance.forceUpdate()

        lastBlock.find('button:first-child').simulate('click')

        expect(spy).toHaveBeenCalled()
      })

      it("should call cancel method when last button is clicked if it's lastBlock with correct user data", () => {
        const spy = jest.spyOn(newInstance, 'cancel').mockImplementation(jest.fn())
        newInstance.forceUpdate()

        lastBlock.find('button:last-child').simulate('click')

        expect(spy).toHaveBeenCalled()
      })

      it("should call cancel method when button is clicked if it's lastBlock with error", () => {
        const errorBlock = shallow(
          <SignUpBlock {...mockProps}
            id={4}
            process={{name: 'signup', step: 4}} />
        )

        const spy = jest.spyOn(errorBlock.instance(), 'cancel').mockImplementation(jest.fn())
        errorBlock.instance().forceUpdate()

        errorBlock.find('button').simulate('click')

        expect(spy).toHaveBeenCalled()
      })
    })
  })

  describe("Lifecycle methods", () => {

    it("should call checkType fn after rendering if there is process prop", () => {
      const spy = jest.spyOn(instance, 'checkType').mockImplementation(jest.fn())
      instance.forceUpdate()

      instance.componentDidMount()

      expect(spy).toHaveBeenCalled()
    })

    it("should call checkType fn if process step prop was changed", () => {
      const spy = jest.spyOn(instance, 'checkType').mockImplementation(jest.fn())
      instance.forceUpdate()

      const mockPrevProps = {
        process: {
          name: 'signup',
          step: 2
        }
      }

      instance.componentDidUpdate(mockPrevProps)

      expect(spy).toHaveBeenCalled()
    })
  })

  describe("checkType", () => {

    it("should change type state to active if process step and id are the same when checkType is called", () => {
      block.setState({type: 'inactive'})
      expect(block.state('type')).toEqual('inactive')
      block.instance().checkType()
      expect(block.state('type')).toEqual('active')
    })

    it("should change type state to inactive if process step is less than id when checkType is called", () => {
      const block = shallow(
        <SignUpBlock {...mockProps}
          process={{
            name: 'signup',
            step: 1
          }}
          id={ 2 }/>
      )
      block.setState({type: 'active'})
      expect(block.state('type')).toEqual('active')
      block.instance().checkType()
      expect(block.state('type')).toEqual('inactive')
    })

    it("should change type state to done if process step is more than id when checkType is called", () => {
      const block = shallow(
        <SignUpBlock {...mockProps}
          process={{
            name: 'signup',
            step: 3
          }}
          id={ 2 }/>
      )
      block.setState({type: 'active'})
      expect(block.state('type')).toEqual('active')
      block.instance().checkType()
      expect(block.state('type')).toEqual('done')
    })
  })

  it("should change data state if handleChanges is called", () => {
    expect(block.state('data')).toEqual({ })

    const mockEvent = {
      target: {
        id: 'username',
        value: 'some-user'
      }
    }

    block.instance().handleChanges(mockEvent)
    expect(block.state('data')).toEqual({'username': 'some-user'})
  })

  describe("updateUser", () => {

    it("should call increaseStep prop when updateUser is called", () => {
      instance.updateUser()
      expect(increaseStep).toHaveBeenCalled()
    })

    it("should call addUserInfo with data state prop when updateUser is called", () => {
      instance.updateUser()
      expect(addUserInfo).toHaveBeenCalledWith(block.state('data'))
    })
  })

  describe("submitUser", () => {

    it("should call addItem with correct argument when submitUser is called", () => {
      addItem.mockImplementation(jest.fn())
      instance.submitUser()
      expect(addItem).toHaveBeenCalledWith('user', {})
    })

    it("should call addItemToAll with correct argument when submitUser is called", () => {
      addItemToAll.mockImplementation(jest.fn())
      instance.submitUser()
      expect(addItemToAll).toHaveBeenCalledWith('users', {})
    })
  })
})

describe("mapStateToProps", () => {

  it("should return the object with correct key", () => {
    const mockStore = {
      process: null,
      user: null,
      error: 'There is error!'
    }

    const expected = {
      process: null,
      user: null,
    }

    const result = mapStateToProps(mockStore)

    expect(result).toEqual(expected)
  })
})

describe("mapDispatchToProps", () => {
  let mockDispatch, mockProps

  beforeEach(() => {
    mockDispatch = jest.fn()
    mockProps = mapDispatchToProps(mockDispatch)
  })

  it("should call dispatch with an addUserInfo action when addUserInfo is called", () => {
    const user = { name: 'Ray' }

    const actionToDispatch = addUserInfo(user)

    const result = mockProps.addUserInfo(user)

    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
  })

  it("should call dispatch with an increaseStep action when increaseStep is called", () => {
    const actionToDispatch = increaseStep()

    const result = mockProps.increaseStep()

    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
  })
})
