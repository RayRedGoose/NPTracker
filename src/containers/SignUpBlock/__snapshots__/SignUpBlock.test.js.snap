// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUpBlock Snapshots should match snapshot if type is active on last step of process and all user data was entered 1`] = `
<section
  className="signup-container active-block"
  id={4}
>
  <div
    className="submit-block"
    key="submit-block"
  >
    <h2>
      Profile overview
    </h2>
    <div
      className="image"
      style={
        Object {
          "backgroundImage": "url(some-image)",
        }
      }
    />
    <p>
      Ray Zlou
    </p>
    <p
      className="location"
    >
      Denver
    </p>
    <div>
      <button
        onClick={[Function]}
      >
        Create
      </button>
      <button
        onClick={[Function]}
      >
        Cancel
      </button>
    </div>
  </div>
</section>
`;

exports[`SignUpBlock Snapshots should match snapshot if type is done 1`] = `
<section
  className="signup-container done-block"
  id={1}
/>
`;

exports[`SignUpBlock Snapshots should match snapshot if type is inactive 1`] = `
<section
  className="signup-container inactive-block"
  id={1}
>
  <img
    alt="matryoshka"
    src="matryoshka.svg"
    style={
      Object {
        "height": "160px",
        "position": "relative",
        "zIndex": "3",
      }
    }
  />
</section>
`;

exports[`SignUpBlock Snapshots should match snapshot with all data passed correctly 1`] = `
<section
  className="signup-container active-block"
  id={1}
>
  <form>
    <h2>
      CREATE ACCOUNT
    </h2>
    <label
      htmlFor="username"
      key="label-username"
    >
      username
    </label>
    <input
      id="username"
      key="input-username"
      onChange={[Function]}
      type="text"
    />
    <p>
      click arrow button to go to next step
    </p>
  </form>
  <img
    alt="arrow"
    className="arrow-btn"
    onClick={[Function]}
    src="send.svg"
  />
</section>
`;

exports[`SignUpBlock Snapshots should match snapshot with error if type is active on last step of process but no user data 1`] = `
<section
  className="signup-container active-block"
  id={4}
>
  <div
    className="submit-block error-block"
    key="error-block"
  >
    <h2>
      Profile overview
    </h2>
    <p
      className="error"
    >
      You don't pass correct data
    </p>
    <button
      onClick={[Function]}
    >
      Cancel
    </button>
  </div>
</section>
`;
